### PR TITLE
Try running html5lib tests efficiently.

### PR DESCRIPTION
**Pull request by desrosj**
Trac ticket: https://core.trac.wordpress.org/ticket/61209.

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by dmsnell** [comment]
@sirreal check out this awesome work @desrosj is doing - runs the html5lib tests in a separate job and _only_ once, instead of on every version of PHP and MySQL. should make it nicer to have the full test suite run on every change without holding up the CI process.

**Comment by sirreal** [comment]
This is great! At the moment, it appears to be running the tests on this branch, is that expected?

I expected it _not to run_ because there are no changes to the relevant files.

**Comment by desrosj** [comment]
It's running for this PR because I included the two relevant GitHub Action workflow files in the list that should run the tests when updated. If there are changes to the PHPUnit workflows, it's possible they could change how the `html5lib` tests are run, so we should run them just to confirm.

File: .github/workflows/phpunit-tests.yml
Comment (sirreal) [inline .github/workflows/phpunit-tests.yml:39]: This needs to be updated before landing.

File: .github/workflows/phpunit-tests.yml
Comment (sirreal) [inline .github/workflows/phpunit-tests.yml:120]: This needs to be updated before landing.

**Review by sirreal — APPROVED** [review]
This seems good, thanks!

**Comment by dmsnell** [comment]
Can we not go ahead and run the tests on changes to every file in Core? there _are_ connections through the `WP_Token_Map` and `kses.php`. The test suite is quite fast, executing in 1.44s on my laptop, including the PHPUnit setup time.

Seems like this wouldn't be much of a cost on the CI runners or delay PRs but could provide important early warnings that something unintentionally changed the parser.

**Comment by desrosj** [comment]
> an we not go ahead and run the tests on changes to every file in Core?

The [test group should only run if the following files are changed](https://github.com/WordPress/wordpress-develop/pull/6963/files#diff-6bd0de9bea001237170fd823cca768f953e6b7286418530e8c44f2ff9e5a535eR188-R202) or patterns matched:
- src/wp-includes/html-api/**.php
- tests/phpunit/tests/html-api/wpHtmlProcessorHtml5lib.php
- tests/phpunit/data/html5lib-tests/**.dat
- .github/workflows/phpunit-tests.yml
- .github/workflows/reusable-phpunit-tests-v3.yml

Any other time the workflow runs the test group will not be run. They are running for this PR because the last two files in that list are being updated to support testing that specific group separately. Are you seeing something indicating differently @dmsnell?

**Comment by dmsnell** [comment]
@desrosj exactly. my point is that I don't understand why we're limiting it to running only when files in the `wp-includes/html-api/` directory change. For one, other files in Core can affect the parser behavior, for two, the tests are really cheap to run. Is there a compelling reason to avoid running them on changes to any Core files?

**Comment by desrosj** [comment]
Ah! Apologies, I misread your comment. This was based on the description on the Trac ticket and the original attempt to solve this in #6546.

@jorbin do you recall any context to that detail?

**Comment by sirreal** [comment]
The reason these tests were excluded from the main test run is [mentioned on the ticket](https://core.trac.wordpress.org/ticket/61209):

> The html-api-html5lib-tests group of tests is excluded by default since it contains many intentionally and dynamically skipped tests (see discussion in #60227 for why this is the best route for these tests).

These tests currently skip 568 tests:

```
OK, but incomplete, skipped, or risky tests!
Tests: 1498, Assertions: 930, Skipped: 568.
```

The amount of skipped tests were some of the main concerns in the original work: https://core.trac.wordpress.org/ticket/60227

The skipped tests may come down a bit more and at some point we may be able to hardcode lists of tests that we do not plan to support to omit them entirely (without skipping).

We can always revisit the decision and consider running these tests along with the rest of the tests.

**Comment by sirreal** [comment]
I'd be very happy to see this move ahead, is there any way I can help?

Anecdotally, some (minor) failures in the html5lib tests slipped into trunk recently. It wasn't a big problem and has already been fixed, but this work would have prevented that from happening.

**Comment by desrosj** [comment]
It looks like today there are `421` skipped tests. Great work getting that number way down, everyone! The total counts for reference are `Tests: 1508, Assertions: 1087, Skipped: 421.`.

I've gone and opened #7602, which is an alternative approach. [Core-59251](https://core.trac.wordpress.org/changeset/59251) added the ability to run only a specific test group, so the new PR takes advantage of that. I think I like the simplicity of that approach instead with the trade off of running the tests more frequently. With the numbers above, the group is now at a ~70% run rate.

**Comment by desrosj** [comment]
Going to close this out in favor of #7602.

**Pull request closed without merging**
